src/hve/arch/intel_x64/external_interrupt.cpp

```
@@ -0,0 +1,101 @@
+{
+
+external_interrupt::external_interrupt(gsl::not_null<eapis::intel_x64::vcpu *> vcpu)
+{
+    using namespace vmcs_n;
+
+    vcpu->exit_handler()->add_handler(
+        exit_reason::basic_exit_reason::external_interrupt,
+        ::handler_delegate_t::create<external_interrupt, &external_interrupt::handle>(this)
+    );
+}
+
+external_interrupt::~external_interrupt()
+{
+    if (!ndebug && m_log_enabled) {
+        dump_log();
+    }
+}
+
+void
+external_interrupt::add_handler(
+    vmcs_n::value_type vector, handler_delegate_t &&d)
+{ m_handlers[vector].push_front(std::move(d)); }
+
+void
+external_interrupt::enable_exiting()
+{
+    vmcs_n::pin_based_vm_execution_controls::external_interrupt_exiting::enable();
+    vmcs_n::vm_exit_controls::acknowledge_interrupt_on_exit::enable();
+}
+
+// -----------------------------------------------------------------------------
+// Debug
+// -----------------------------------------------------------------------------
+
+void
+external_interrupt::dump_log()
+{
+    bfdebug_transaction(0, [&](std::string * msg) {
+        bfdebug_lnbr(0, msg);
+        bfdebug_info(0, "External interrupt counts", msg);
+        bfdebug_brk2(0, msg);
+
+        for (auto i = 0U; i < 256U; ++i) {
+            if (m_log[i] > 0U) {
+                bfdebug_subnhex(0, std::to_string(i).c_str(), m_log[i], msg);
+            }
+        }
+
+        bfdebug_lnbr(0, msg);
+    });
+}
+
+// -----------------------------------------------------------------------------
+// Basic exit reason handler
+// -----------------------------------------------------------------------------
+
+bool
+external_interrupt::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    info_t info{vmcs_n::vm_exit_interruption_information::vector::get()};
+
+    if (!ndebug && m_log_enabled) {
+        m_log[info.vector]++;
+    }
+
+    for (const auto &d : m_handlers[info.vector]) {
+        if (d(vmcs, info)) {
+            return true;
+        }
+    }
+
+    return false;
```

Returning false here is a bad idea as that will result in an unhandled VM exit error as the base hypervisor doesn't support this type of exit. Likely this should just throw a useful error here.